### DISCUSSION TITLE
Avalonia Rendering Quirks

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#9233] [General]
## Avalonia Rendering Quirks

I didn't want to make an issue for this as it's unclear which direction things will go. However, I've noticed a few fundamental rendering differences Avalonia has compared to WPF or WinUI and I think these should be discussed before 11.0 release. Consider this a roll-up or epic encompassing all rendering-related differences.

 1. `CornerRadius` Clipping (compared to WinUI)
     * The `CornerRadius` property comes from WinUI (formerly UWP) and there setting the corner radius of a parent will automatically clip the children to the corner radius. There is no need to set the radius separately on the children -- or use a clipping geometry. In Avalonia, the parent does NOT clip its children in this situation which is unexpected and 
     * See the issue https://github.com/AvaloniaUI/Avalonia/issues/2105
     * Note that this behaves as WPF would in this situation. However, WPF is unintuitive to work with in this case and the `CornerRadius` was really integrated in UWP which should be the target for this feature.
 3. Missing `Control.BackgroundSizing` (compared to WinUI)
     * [`Control.BackgroundSizing`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.control.backgroundsizing?view=winrt-22621)
     * [`BackgroundSizing enum`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.backgroundsizing?view=winrt-22621)
     * In WinUI `BackgroundSizing` is used to control how the background is rendered relative to any border. This gets really important when borders are semi-transparent.
     * Avalonia right now draws the background to the middle of the border stroke -- Skia's default. However, that mode causes more trouble and isn't even an option in WinUI.
     * This is partially implemented in FluentAvalonia here: https://github.com/amwx/FluentAvalonia/pull/211#issuecomment-1281506805
 3. `Opacity` (compared to WPF/WinUI) 
     * This is best described in the new issue: https://github.com/AvaloniaUI/Avalonia/issues/9227
     * Basically, Avalonia child controls inherit the Opacity property of the parent. Then this property is applied to each child separately. However, this is significantly different from WPF which essentially creates a union of the rendered children and then applies opacity once to the union.

The point of discussion is to figure out how these should be handled in the future. After 11.0 changes to address this sort of thing might be very difficult so it needs to be talked-about before.

**Comment by maxkatz6** [comment]
> There is no need to set the radius separately on the children -- or use a clipping geometry. In Avalonia, the parent does NOT clip its children in this situation which is unexpected and

It does clip it, if ClipToBounds is set.
Linked issue is about problem when children overlap the border. 

> Missing Control.BackgroundSizing (compared to WinUI)

Yes, this one I am missing too. Not sure about others opinion on that.

> Opacity (compared to WPF/WinUI)

I think it was planned to be fixed, but nobody got to it yet.

**Reply by robloo** [reply]
> It does clip it, if ClipToBounds is set.
> Linked issue is about problem when children overlap the border.

Well, I know we've talked about this a few different times but I'm still not understanding you. I ran across this for sure with DropDownButton and you can still see it today in the ControlCatalog.

![image](https://user-images.githubusercontent.com/17993847/196569448-f0ce2271-816d-42fb-81ec-9ac8cca6102b.png)

And the source code of the root border with the corner radius has ClipToBounds set:

https://github.com/AvaloniaUI/Avalonia/blob/f7f17224bb5684136cbd62a9871d73d3db03f40a/src/Avalonia.Themes.Fluent/Controls/DropDownButton.xaml#L30-L35

What am I missing? I couldn't fix this with ColorPicker and just ended up working around the issue by setting the `CornerRadius` of the child as well. We talked about that in the PR briefly.

Edit: So basically you are saying the Control property works fine. The issue is with the `Border` only?

**Reply by kekekeks** [reply]
There is a regression with compositing renderer that has broken round corner clipping, check if it works with the old one.

**Reply by robloo** [reply]
DropDownButton was implemented towards the beginning of the year -- several months before the compositing renderer was merged. I have noticed this issue since the beginning.

If I have a chance I'll double check with the old deferred renderer though.

**Reply by timunie** [reply]
I think there is another thing. Clipping doesn't respect the border thickness. That way the content overlays the border which looks wired.

**Reply by robloo** [reply]
@timunie Can you give a bit more details for that issue? I'll add it to the discussion list itself.

**Reply by timunie** [reply]
Sure. I think it's related to this issue https://github.com/AvaloniaUI/Avalonia/issues/2105#issuecomment-888617422 . You may notice it better if you apply a bigger radius and a thicker border brush. 

I can't double-check atm for new renderer as I have no PC at hand

**Comment by robloo** [comment]
Concerning `Control.BackgroundSizing`, I think this can be added at any point in the future without many bad breaking changes. It will only slightly re-work how some of the default ControlThemes work internally. UWP also added this later-on I believe.

However, in order to keep compatibility I think Avalonia will need to:
 1. Support an additional option like `CenterBorder` in addition to the WinUI values `InnerBorderEdge` and `OuterBorderEdge`. `CenterBorder` represents how Avalonia currently works and would be the default for compatibility in the future.
 2. Switch now to using either `InnerBorderEdge` or `OuterBorderEdge`. That would be a bit cleaner and we wouldn't need a 3rd value for compatibility in the future. Rendering the background to the center of the border also doesn't currently work well with transparent border brushes.
